pkg/daemon/daemon.go

```diff
@@ -484,35 +485,34 @@ func ReexecuteForTargetRoot(target string) error {
 	targetBinBase := "run/bin/machine-config-daemon"
 	targetBin := filepath.Join(target, targetBinBase)
 
-	// Be idempotent
-	if _, err := os.Stat(targetBin); err != nil {
-		sourceBinary := "/usr/bin/machine-config-daemon" + sourceBinarySuffix
-		src, err := os.Open(sourceBinary)
-		if err != nil {
-			return fmt.Errorf("opening %s: %w", sourceBinary, err)
-		}
-		defer src.Close()
+	targetBinDir := filepath.Dir(targetBin)
+	if err := os.MkdirAll(targetBinDir, 0o755); err != nil {
+		return fmt.Errorf("mkdir %s: %w", targetBinDir, err)
+	}
 
-		targetBinDir := filepath.Dir(targetBin)
-		if _, err := os.Stat(targetBinDir); err != nil {
-			if err := os.Mkdir(targetBinDir, 0o755); err != nil {
-				return fmt.Errorf("mkdir %s: %w", targetBinDir, err)
-			}
-		}
+	f, err := renameio.TempFile(targetBinDir, targetBin)
+	if err != nil {
+		return err
+	}
+	defer f.Cleanup()
```

Comment:
Yes I believe we still need this to close the file in case of error (not that that matters much, but still).

It's the pattern used in `writeFileAtomically()` too.  Looking at the source, this is how it's intended to be used - there's an internal `done` variable and `Cleanup` is correctly a no-op if `writeFileAtomically()` succeeds.